[XLA:GPU][oneAPI] SYCL memcpy functions and tests

bhavani-subramanian opened the pull request:
Co-author: @kanvi-nervana 

This PR implements SYCL memcpy functions and adds corresponding tests. It also implements some misc. features in `sycl_gpu_runtime` such as functions to get SYCL frequency, synchronize stream etc.

It depends on https://github.com/openxla/xla/pull/32918.

bhavani-subramanian committed:
Added SyclStreamPool and tests

bhavani-subramanian committed:
Added SYCL memory management functions and tests

bhavani-subramanian committed:
Added SYCL memcpy functions and tests

beckerhe commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -107,4 +196,514 @@ absl::StatusOr<::sycl::device> SyclDevicePool::GetDevice(int device_ordinal) {
+                     << ".";
+        }
+      }
+    };
```
This define a global object of a non-trivial type (std::function). This is discouraged in the Google C++ style guide because it can lead to initialization and destruction order issues.

Since the lambda is stateless, it should be possible to just declare a function here:

```
void SyclAsyncHandler(::sycl::exception_list ex_list) {
// ...
}
```

The conversion into to a std::function will then happen when the function is getting passed to the sycl::queue constructor.

beckerhe commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -107,4 +196,514 @@ absl::StatusOr<::sycl::device> SyclDevicePool::GetDevice(int device_ordinal) {
+  try {
+    // Use the default stream to allocate memory
+    void* ptr =
+        aligned_alloc_shared(/*alignment=*/64, byte_count, *stream_handle);
```
This is sycl::aligned_alloc_shared, right? Can you qualify this call?

beckerhe commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -107,4 +196,514 @@ absl::StatusOr<::sycl::device> SyclDevicePool::GetDevice(int device_ordinal) {
+  try {
+    // Use the default stream to allocate memory
+    void* ptr =
+        aligned_alloc_host(/*alignment=*/64, byte_count, *stream_handle);
```
This is sycl::aligned_alloc_host, right? Can you qualify this call?

bhavani-subramanian replied:
Fixed.

beckerhe commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -107,4 +196,514 @@ absl::StatusOr<::sycl::device> SyclDevicePool::GetDevice(int device_ordinal) {
+    return absl::OkStatus();
+  }
+  ::sycl::usm::alloc dst_alloc_type =
+      get_pointer_type(dst_host, stream_handle->get_context());
```
Can you add `sycl::` to this function call?

bhavani-subramanian replied:
Done.

beckerhe commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -107,4 +196,514 @@ absl::StatusOr<::sycl::device> SyclDevicePool::GetDevice(int device_ordinal) {
+            << device_ordinal << ", stream pool size is "
+            << stream_pool->size();
+    return absl::OkStatus();
+  }
```
We try to avoid `else` blocks after `return` statements.

bhavani-subramanian replied:
Fixed.

beckerhe commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -107,4 +196,514 @@ absl::StatusOr<::sycl::device> SyclDevicePool::GetDevice(int device_ordinal) {
+  return MemsetDevice(stream_handle.get(), dst_device, value, count);
+}
+
+absl::Status SyclMemsetDeviceAsync(::sycl::queue* stream_handle,
```
Can you either make `stream_handle` a reference instead of a pointer (preferred) or add the `absl_nonnull` attribute?

bhavani-subramanian replied:
Done.

beckerhe commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -107,4 +196,514 @@ absl::StatusOr<::sycl::device> SyclDevicePool::GetDevice(int device_ordinal) {
+  return MemfillDevice(stream_handle.get(), dst_device, value, count);
+}
+
+absl::Status SyclMemfillDeviceAsync(::sycl::queue* stream_handle,
```
Can you either make `stream_handle` a reference instead of a pointer (preferred) or add the `absl_nonnull` attribute?

bhavani-subramanian replied:
Done.

beckerhe commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -38,6 +37,96 @@ absl::Status IsValidDeviceOrdinal(int device_ordinal,
+  return absl::OkStatus();
+}
+
+absl::Status MemfillDevice(::sycl::queue* stream_handle, void* dst_device,
```
Can you either make `stream_handle` a reference instead of a pointer (preferred) or add the `absl_nonnull` attribute?

bhavani-subramanian replied:
Done.

beckerhe commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -38,6 +37,96 @@ absl::Status IsValidDeviceOrdinal(int device_ordinal,
+  return absl::OkStatus();
+}
+
+absl::Status MemcpyHostToDevice(::sycl::queue* stream_handle, void* dst_device,
```
Can you either make `stream_handle` a reference instead of a pointer (preferred) or add the `absl_nonnull` attribute?

beckerhe reviewed (commented):
Thank you for this change. This looks overall good to me. My comments are mainly just style things.

bhavani-subramanian committed:
Addressed review comments:
- Use a function for SyclAsyncHandler
- Qualified all SYCL functions with ::sycl::
- Use absl::flat_hash_map for StreamPoolMap
- Removed `else` blocks after `return` statements
- Added ABSL_ATTRIBUTE_NONNULL for stream_handle
- Use TF_ASSERT_OK instead of ASSERT_TRUE(...ok())
- Use EXPECT_THAT instead of EXPECT_EQ when checking error status

bhavani-subramanian commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -107,4 +196,514 @@ absl::StatusOr<::sycl::device> SyclDevicePool::GetDevice(int device_ordinal) {
+  try {
+    // Use the default stream to allocate memory
+    void* ptr =
+        aligned_alloc_device(/*alignment=*/64, byte_count, *stream_handle);
```
Fixed.

bhavani-subramanian commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -107,4 +196,514 @@ absl::StatusOr<::sycl::device> SyclDevicePool::GetDevice(int device_ordinal) {
+                              byte_count);
+}
+
+absl::Status SyclMemcpyDeviceToHostAsync(::sycl::queue* stream_handle,
```
Done.

bhavani-subramanian commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -38,6 +37,96 @@ absl::Status IsValidDeviceOrdinal(int device_ordinal,
+  return absl::OkStatus();
+}
+
+absl::Status MemcpyDeviceToDevice(::sycl::queue* stream_handle,
```
Done.

bhavani-subramanian commented on xla/stream_executor/sycl/sycl_gpu_runtime.cc:
```diff
@@ -38,6 +37,96 @@ absl::Status IsValidDeviceOrdinal(int device_ordinal,
+  return (__LIBSYCL_MAJOR_VERSION >= 7) && (__LIBSYCL_MINOR_VERSION >= 2);
+}
+
+absl::Status MemcpyDeviceToHost(::sycl::queue* stream_handle, void* dst_host,
```
Done.

bhavani-subramanian commented on xla/stream_executor/sycl/sycl_gpu_runtime_test.cc:
```diff
@@ -43,5 +126,580 @@ TEST(SyclGpuRuntimeTest, TestStaticDeviceContext) {
+                                              /*enable_multiple_streams=*/false)
+                .status()
+                .code(),
+            absl::StatusCode::kInvalidArgument);
```
Fixed.

I used `absl_testing::StatusIs(...)` instead of `tsl::testing::StatusIs(...)` since I get `warning: 'StatusIs<absl::StatusCode>' is deprecated`.
